### DISCUSSION TITLE
From semi controlled elements to controlled nodes and edges

### DISCUSSION THREAD

**Discussion by moklick** [xyflow/xyflow#960] [Ideas]
# From elements to nodes and edges

At the moment there is a lot of uncertainty about handling elements (nodes and edges). The problem is that React Flow stores and updates some values like the position, width, height and handles internally that are not changed in the elements array that you pass to React Flow. It would be much clearer if we would use a controlled `elements` prop in combination with an `onElementsChange` handler to pass the changes (dragging, selecting, removing) to your elements array. 

Now that we are speaking about breaking changes I thought that it would be a good time to separate nodes and edges and don't use the elements array anymore but `nodes` and `edges` with `onNodesChange` and `onEdgesChange` handlers.

That's the idea for now. We will think about it and discuss pros, cons and the specific api changes here.

**Comment by mikebarkmin** [comment]
I think that it is a good idea to have separate props for nodes and edges. At least for my use-case I am storing nodes and edges separatly and combine them to one array for passing it to the props of the flow.

**Reply by ambroseus** [reply]
+1 my case too
![Selection_012](https://user-images.githubusercontent.com/380645/112794744-0cc66e80-9070-11eb-902f-4c0f34c9022b.jpg)

**Comment by computebender** [comment]
This would be amazing for my use case. I'm using React Flow as a part of a larger app where nodes mean something beyond just on the canvas. In our use case clicking a node lets you configure that node somewhere else. To manage that, I'm attempting to store the source of truth of all nodes outside React Flow's state. This means when a node is added to the external state, a new node is added to React Flow and when a node in React Flow gets removed, moved, updated, the external state must be updated to match. 

The "syncing" of states I'm trying to do would be much easier with callbacks just representing element changes. An example of where the current implementation falters is having multiple nodes selected and moved, I'm relying on the onNodeDragStop callback to update external state when a node is moved, but this doesn't work for multiple selected nodes (as the callback is only called on one), or probably for "Auto-Aligning" nodes (haven't tried, but don't see why the drag callback would be called in that case).

**Comment by meatflavourdev** [comment]
> # From elements to nodes and edges
> At the moment there is a lot of uncertainty about handling elements (nodes and edges). The problem is that React Flow stores and updates some values like the position, width, height and handles internally that are not changed in the elements array that you pass to React Flow. It would be much clearer if we would use a controlled `elements` prop in combination with an `onElementsChange` handler to pass the changes (dragging, selecting, removing) to your elements array.

+1 for this for sure-- This is the main limitation I've had to contending with-- that there is internal data hidden away inside redux stores and the `elements` props isn't controlled so there is no 1:1 relationship between the exposed elements state array and the state of the elements as React Flow sees them. Furthermore, when integrating with 3rd party libraries that need to synchronize state with React Flow, it's a headache that there isn't an easy solution to:

a) Give me an easy way to prevent React Flow from updating position on node drag (getting weird race conditions when I attempt to update the position myself on node drag)
b) Give me an easy way to observe React Flow's state changes  without subscribing to every redux update and filtering through each update frame myself (is there a better way to do this? 🤷‍♂️) 

Your proposal would increase the interoperability of the library and I think that's good for the library in the long run-- even if the change is not backwards compatible.

As for an API.. something like this I would assume-- forgive any glaring errors, I'm quite tired.
```javascript
export type elementChange<T = any> = {
  before: FlowElement, // Before element state
  after: FlowElement, // After element state
  changedKeys: Map<string, any> // A map from each changed key to its new value, for ease of access
};

ReactFlowProps.onElementsChange?: (changedElements: elementChange[]) => void;
```

**Reply by meatflavourdev** [reply]
### Thoughts on considerations re: upcoming changes
It seems to me, this would be an ideal time to also think of grouping and rafting-- 
check out this [great example of a raft composite](http://www.draw2d.org/draw2d_touch/jsdoc_6/#!/example/composite_raft)
and a [more common group composite](http://www.draw2d.org/draw2d_touch/jsdoc_6/#!/example/composite_group)
Both would be excellent additions-- but my point is more that they're worth keeping in mind when considering modification to the data structure, as potential futures that present decisions might be aware of.

In addition to that-- the concept of visual ordering might also be kept in mind? Consider a scene graph where layers can be rearranged. Just throwin' all that out there as something to consider 🤷‍♂️😅

**Comment by bentron2000** [comment]
I imagine that one of the newer state libs like recoil/jotai/hookstate will make it fairly straightforward for you to provide hooks to subscribe to / update internal state more easily.

**Comment by lostfictions** [comment]
Would love to help make the transition to controlled nodes and edges happen! Maybe it could be opt-in and shipped experimentally for now, so that it's not a huge breaking change and can be iterated on? (ie. the same way React form elements can be controlled or uncontrolled)

**Reply by meatflavourdev** [reply]
+1 hookstate
ActiveJS is also just great as it implements the Observable contract for Rxjs inoperability-- a boon for anyone like me who loves functional reactive programming; any updates can be subscribed to implicitly taking care of notifying whoever and whatever wants to be notified.

yJS would fit right in by design as a [**Custom AsyncSystem** ](https://docs.activejs.dev/fundamentals/systems/custom-asyncsystem) ~~Whoops wrong thread...~~ but my comment stands, [yJS](https://github.com/yjs/yjs) whips the Llamas a**.

**Comment by moklick** [comment]
I am currently refactoring the node and edge handling and the internal/external state handling. The current API looks like this: 

```js 
const [nodes, setNodes] = useState<Node[]>(initialNodes);
const [edges, setEdges] = useState<Edge[]>(initialEdges);

const onNodesChange = useCallback((changes: ElementChange[]) => {
  setNodes((prevNodes) => applyNodeChanges(changes, prevNodes));
}, []);

const onEdgesChange = useCallback((changes: ElementChange[]) => {
  setEdges((prevEdges) => applyEdgeChanges(changes, prevEdges));
}, []);

return (
  <ReactFlow
    nodes={nodes}
    edges={edges}
    onNodesChange={onNodesChange}
    onEdgesChange={onEdgesChange}
  />
);
```

With this change you get notified when a node (or multiple nodes) get dragged, selected or removed. The `applyNodeChanges` and `applyEdgeChanges` functions get exported from the library. What do you think about this API?

**Reply by mikebarkmin** [reply]
I assume that ElementChange is defined something likes this:
```typescript
type ElementChange = {
  event: "dragged" | "selected" | "removed",
  element: Node | Edge
}
```

I think it would be better to have EdgeChange and NodeChange as individual types:

```typescript
type EdgeChange = {
  type: "edge";
  event: "selected" | "removed";
  element: Edge
}

type NodeChange = {
  type: "node"
  event: "dragged" | "selected" | "removed";
  element: Node;
}

type ElementChange = EdgeChange | NodeChange;
```

Then the `onNodesChange` and `onEdgesChange` can be defined as:

```typescript
type OnNodesChange = (changes: NodeChange[]) => void;
type OnEdgesChange = (changes: EdgeChange[]) => void;
```

**Reply by mikebarkmin** [reply]
Another though I had is that we could introduce a custom hook:
```typescript
const [nodes, applyNodeChanges, setNodes] = useNodeState(initialNodes);
const [edges, applyEdgeChanges, setEdges] = useEdgeState(initialEdges);

return (
  <ReactFlow
    nodes={nodes}
    edges={edges}
    onNodesChange={applyNodeChanges}
    onEdgesChange={applyEdgeChanges}
  />
);
```

**Reply by moklick** [reply]
Thanks! For now `ElementChange` is defined like this: 

```ts 
export type ElementChange = {
  id: string;
  change?: any;
  delete?: boolean;
};
```

if a node or a edge gets selected the change look like this:

```js
{ id: 'id-123', change: { selected: true } }
```

if a node gets dragged the change look like this:

```js
{ id: 'id-123', change: { position: { x: 100, y: 200 } } }
```

I like the idea of exporting a help hook!

**Reply by lostfictions** [reply]
Hey @moklick, thanks for the update! I strongly agree with @mikebarkmin that it should be two different signatures for the two different change callbacks -- that'll make type inference and usability much better right off the bat. Is there currently a good reason for them to be combined together into a single type, besides the fact that that was how it worked previously? As @mikebarkmin notes, even if you need a unified event type, you can represent it as something like this:

```ts
type ElementChangeEvent = EdgeChangeEvent | NodeChangeEvent;
```

---

Regarding the draft API:

> if a node or a edge gets selected the change look like this:
> 
> ```js
> { id: 'id-123', change: { selected: true } }
> ```
> 
> if a node gets dragged the change look like this:
> 
> ```js
> { id: 'id-123', change: { position: { x: 100, y: 200 } } }
> ```

Personally, I'd be strongly in favor of some kind of discriminant property to make Typescript inference work better. As it stands with your proposal, it would theoretically be possible to discriminate the type of change by doing something like this:

```ts
type ElementChange = { id: string } & (
 | { change: { selected: boolean } }
 | { change: { position: Position } }
)

function handleChange(change: ElementChange) {
  if('position' in change.change) {
    // in this block TS can infer this is a drag event and
    // narrow the type of `change.change` appropriately
  }
  else if('selected' in change.change) {
    // ditto for selection events
  }
  // ...and so on...
}
```

...but this forces consumers of the library to write relatively ad-hoc and complex code to narrow the type of the change. It's also somewhat brittle if you introduce more complex change types. Instead, I'd recommend using [tagged union types](https://mariusschulz.com/blog/tagged-union-types-in-typescript), which Typescript is very good at narrowing. That might look something like this:

```ts
type NodeChangeEvent = { id: string } & (
  | { type: "move", position: Position }
  | { type: "select", selected: boolean }
  | { type: "delete" }
)
```

If you prefer, those individual event types can also be defined separately, for example:

```ts
type NodeMoveEvent = {
  id: string;
  type: "move";
  position: Position;
}

type NodeSelectEvent = {
  id: string;
  type: "select";
  selected: boolean;
}

type NodeDeleteEvent = {
  id: string;
  type: "delete";
}

type NodeChangeEvent =
  | NodeMoveEvent
  | NodeSelectEvent
  | NodeDeleteEvent
```

...or the common parts (the `id` in this case) can be factored out to the top-level union type but the sub-event types can still be named, ie. `type NodeChangeEvent = { id: string } & (NodeMoveEvent | NodeSelectEvent | NodeDeleteEvent)`.

Regardless, consumers could then consume those events like this:

```ts
function onNodeChange(ev: NodeChangeEvent) {
  switch(ev.type) {
    case "move":
      // ev.position is defined in this block!
      console.log(`node moved! ${ev.position}`);
      break;
    case "select":
      // ev.selected is defined here!
      break;

    // etc...
  }
}
```

This also makes it easy to do [exhaustiveness checking](https://www.fullstory.com/blog/discriminated-unions-and-exhaustiveness-checking-in-typescript/) in the `default` block of a switch statement (it works equally well at the end of an if-else chain too, of course). If you as the library author add a new type of event, this then helps ensure that consumers don't accidentally neglect to handle it if they're trying to cover every event type.

---

As an aside, in 2021 there's usually never a reason to use `any` in Typescript to represent a type internal to your codebase. The main use-case for it is at the boundaries of your code, if you're handling some input that could genuinely be anything (for example, arbitrary JSON). But even those cases are rare and are better-served by the [`unknown` type introduced in Typescript 3.0](https://mariusschulz.com/blog/the-unknown-type-in-typescript).

Using `any` is especially dangerous in libraries -- since it basically turns off the typechecker for a given field, it has a tendency to "leak" downstream into consumers' code and cause additional errors there.

**Reply by mikebarkmin** [reply]
Great explanation! This is the way to go in my opinion.

**Reply by moklick** [reply]
This is super helpful @lostfictions :) Thanks! I will separate the node and edge changes and update the API.

**Reply by moklick** [reply]
@mikebarkmin I thought about the hooks. We could add them as helpers but in most cases you won't need them I think. When you are using some state mng lib you would rather export an onNodeChanges action and use that one instead of going the extra step with the local component state. 

@lostfictions @mikebarkmin Would you prefer to get the changes and use the React Flow `applyNodeChanges` and `applyEdgeChanges` helper (A) or to get the nodes/edges with the already applied changes (B)?

A:
```js 
const onNodesChange = useCallback((changes: NodeChange[]) => setNodes((nds) => applyNodeChanges(changes, nds)), []);
const onEdgesChange = useCallback((changes: EdgeChange[]) => setEdges((eds) => applyEdgeChanges(changes, eds)), []);
```

B: 
```js 
const onNodesChange = useCallback((nextNodes: Node[]) => setNodes(nextNodes), []);
const onEdgesChange = useCallback((nextEdges: Edge[]) => setEdges(nextEdges), []);
```

**Reply by datoslabs** [reply]
My apologies in advance if I am asking questions that have already been addressed since I just started to follow this discussion.  Is my understanding correct that with option (A), developers are expected to include boilerplate `onNodesChange` and `onEdgesChange` callbacks in order to affect change?

**Reply by moklick** [reply]
Hey @datoslabs You would need to add these handlers for A and for B. But this is just an example. If you would use a state mng lib the implementation would like differently. This is a good way for us to ensure that the user state and the internal state are synced. Nodes and edges would be fully controlled after this change.

**Reply by lostfictions** [reply]
Hey @moklick, if I understand correctly (B) would give you the entire new state array? I'm not sure that makes any sense, personally -- any consumer who needed fine-grained changes would then need to go back and diff the new state with their current state, which would be really inefficient.

For example, if a consumer is using Immer for state management, they can in theory simply return an entirely new state -- ie. the new node or edge array -- but if they're getting a [patch stream](https://immerjs.github.io/immer/patches) of changes, I think this will consider the entire array changed and create a huge patch even if only one or two elements changed. I believe it's similar for most observable libraries (MobX, etc.), which usually operate at the level of fine-grained changes. Very few state management scenarios would benefit from losing access to change events.

Beyond the inefficiency of making consumers re-extract changes, in general it just feels more idiomatic and common in React to return changes rather than full new states. Is the rationale for (B) that it would cut down on the boilerplate of having to write `useCallback(chs => setNodes(ns => applyNodeChanges(chs, ns)))` in the case where nodes and edges are kept in simple `useState` calls? Personally I don't think it's a huge deal -- it makes it clearer that there's no "magic" happening.

**Reply by moklick** [reply]
Hey @lostfictions @datoslabs @mikebarkmin @meatflavourdev 

If you find the time it would be super helpful if you could check out the current v10 and let me know what you think about the current API. You can read about the changes here: https://github.com/wbkd/react-flow/pull/1555

**Reply by datoslabs** [reply]
THANK YOU @moklick and team for all the amazing work!  I briefly read through the changelog and am impressed with the amount of work that went into this major release!  I will find time this weekend to check out the updates and share my findings/thoughts.

**Reply by lostfictions** [reply]
Thanks for the ping and for all the work on this release @moklick! At a glance it looks great to me -- I'll try to give it a shot in a project this week.

**Reply by moklick** [reply]
Now that we did these changes I am thinking about doing the same for the viewport (x,y,zoom). This would make it easier to fit bounds and doing animated pans/zooms for example. 

This would also mean that we can't have interactive (panning/zooming) flows without defining something like this anymore:

```js 
const [viewport, setViewport, onViewportChange] = useViewportState(initialViewport);
```

🧐

**Reply by mikebarkmin** [reply]
@moklick I am also very impressed. Great explanation of the changes in the upcoming release. These changes will improve the DX of React Flow immensely IMHO. Thank you very much for your afford.

**Comment by datoslabs** [comment]
Hi @moklick,

I didn't get a chance to try out the development version this week; hopefully, I will be able to spend more time next week to gain some hands-on experience on V10 branch.  Based on reviewing the changes, I have a few questions/observations:

A.   Using the `useNodeState` and `useEdgeState` hooks can avoid the need for boilerplate code like:
```  
  const onNodesChange = useCallback((changes) => setNodes((ns) => applyNodeChanges(changes, ns)), []);
  const onEdgesChange = useCallback((changes) => setEdges((es) => applyEdgeChanges(changes, es)), []);
```
The hooks are very convenient, especially for new users, and should be included as the default example.  I am wondering if the hooks can further "store" the default `onNodesChange` and `onEdgesChange` callbacks in the `ReactFlowProvider` context so that if users did not specify/override with custom `onNodesChange` and/or `onEdgesChange` callbacks to the `ReactFlow` component (like the following example), it will use these defaults.  This will further simplify the starting code for new users.
```
...
const BasicFlow = () => {
  const [nodes, setNodes, onNodesChange] = useNodesState(initialNodes);
  const [edges, setEdges, onEdgesChange] = useEdgesState(initialEdges);

  return (
    <ReactFlow
      nodes={nodes} 
      edges={edges}
    />
  );
};

...
```

B.  The `applyChanges` method (in `utils/changes.ts`) requires the `NodeChange[] | EdgeChange[]` array as input; however, the case statements only handle `'select', 'dimensions', and 'remove'.  I may have missed it due to my limited time to look through all the code; how would the user specify other types of changes such as "add" or other updates, are `NodeChange[] | EdgeChange[]` arrays optional (ie just apply the nodes/edges arrays as provided)?

C.   With the overhaul of internal states API, how would the user be able to obtain the `node.__rf.handleBounds` (which is often needed with advanced styling) for auto layout under the new API?

**Reply by moklick** [reply]
Hey @datoslabs 

thanks for your input here :)

A: The helpers are just meant for handling the local state. We now added `defaultNodes` and `defaultEdges` which might lead to the decision to remove the `useNodesState` and `useEdgesState` hooks again ;) There is no real benefit for these hooks anymore. If you want to get started quickly you can use the new uncontrolled behaviour with `defaultNodes` and `defaultEdges` and if you want to have the full control you can use the `nodes`, `edges`, `onNodesChange` and `onEdgesChange` props. 

B: The applyChanges handlers are only for updating changes that are triggered internally by React Flow. If you want to add or alter a node you can just update the node object in your `nodes` array.

C: For now the only way is to get the `nodeInternals` from the store. We already added a new hook `useNodeBounds`. We might add the handles there or create another one that returns the handles.

**Comment by moklick** [comment]
@datoslabs @mikebarkmin @lostfictions @ambroseus @meatflavourdev (I hope it's ok when I ping you here, if not just let me know).

Besides the new way of having an controlled React Flow component we might also add the possibility to create a uncontrolled component by passing `defaultNodes`, `defaultEdges` and `defaultEdgeOptions` . An uncontrolled React Flow component doesn't require any other handlers and everything (drag, remove, connect) gets managed internally. ~~For updating the nodes and edges we created the hooks `useUpdateNodes` and `useUpdateEdges`.~~ 

**Update:** For updating the nodes and edges we will add a `useReactFlow` hook that exports helper functions like `setNodes`, `setEdges`, `addNodes`, `addEdges` etc.
 
You can find an example that uses all the new features here: https://github.com/wbkd/react-flow/blob/v10/example/src/DefaultNodes/index.tsx 

What do you think about this?

**Reply by ambroseus** [reply]
@moklick  looks good :)

**Reply by moklick** [reply]
We will not add separate `useUpdateNodes`, `useUpdateEdges` hooks etc. but add a `useReactFlow` hook that exports functions like `setNodes`, `setEdges`, `addNodes`, `addEdges` etc.

**Reply by lostfictions** [reply]
@moklick that link seems to be broken now -- is it this file here? https://github.com/wbkd/react-flow/blob/2a4dfce61c99f436ed3c4fc5e6fd086995ed2c55/example/src/DefaultNodes/index.tsx

I might be a little confused but it seems like you're dropping the idea of having a controlled component then? This very much looks like an API where you have to call imperative updater functions if your data changes (from an external source, for example). It definitely seems like it would be clunkier, less interoperable, and less useful than the previous proposed API in the case of data being updated from an external source.

Is it possible you accidentally swapped the terms "uncontrolled" and "controlled" in your post? I'm not sure what the "uncontrolled" API is, if it's not this.

**Reply by moklick** [reply]
Thanks! I just corrected the things you mentioned (I always mix up controlled and uncontrolled 🙃) The new uncontrolled API is an extension to the existing one. You can have your controlled flow or you can create an uncontrolled one. The uncontrolled one is easier to setup and good enough for most simpler use cases. If you need the full control you can create a controlled flow with the `onNodesChange` and `onEdgesChange` handlers.

**Reply by datoslabs** [reply]
Thank you @moklick and team for lots of great work and improvements in V10!  The uncontrolled example answered my previous comments about removing the boilerplate code to simplify coding for straight forward use cases as well as users/developers new to React Flow.  

After a quick glance, I noticed the `get*` hooks return shallow copies (instead of deep clone) of the internal store objects, is this by design?  Developers may be able to make changes to these objects (e.g. `node.data`) outside of the `set*` hooks and potentially cause unintended changes in future renderings.  I wonder if we can make these hooks behave much like React `useState` hook with a one-way update after each rendering cycle.
